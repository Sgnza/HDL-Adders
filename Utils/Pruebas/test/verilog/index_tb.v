`include "../../src/verilog/index.v"
module index_tb;

reg clk_tb;
reg [15:0] in_tb;
wire [3:0] out0_tb;
wire [3:0] out1_tb;
wire [3:0] out2_tb;
wire [3:0] out3_tb;

index d0(
    .in(in_tb),
    .out0(out0_tb),
    .out1(out1_tb),
    .out2(out2_tb),
    .out3(out3_tb)
);

task verify;
    input [15:0] in_task;
    input [3:0] out0_task;
    input [3:0] out1_task;
    input [3:0] out2_task;
    input [3:0] out3_task;

    begin
        in_tb = in_task;
        #1;
        if(out0_task == out0_tb && out1_task == out1_tb && out2_task == out2_tb && out3_task == out3_tb) begin
            $display("CORRECTO: in: %b, out0 obtenido: %b, out1 obtenido: %b, out2 obtenido: %b, out3 obtenido: %b", in_tb, out0_tb, out1_tb, out2_tb, out3_tb);
        end else begin
            $display("INCORRECTO: in: %b, out0 obtenido: %b, out1 obtenido: %b, out2 obtenido: %b, out3 obtenido: %b", in_tb, out0_tb, out1_tb, out2_tb, out3_tb);
        end
    end

endtask

initial begin
  $dumpfile("index_tb.vcd");
  $dumpvars(0, index_tb);
end

initial begin

verify(
    16'b1100_0011_1111_1111,
    4'b1111,
    4'b1111,
    4'b0011,
    4'b1100
);

verify(
    16'b1100001111111111,
    4'b0000,
    4'b0000,
    4'b0000,
    4'b0000
);

#100;
$finish;

end

endmodule