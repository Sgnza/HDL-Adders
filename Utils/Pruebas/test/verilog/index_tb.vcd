$date
	Thu Jan 18 23:46:48 2024
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module index_tb $end
$var wire 4 ! out3_tb [3:0] $end
$var wire 4 " out2_tb [3:0] $end
$var wire 4 # out1_tb [3:0] $end
$var wire 4 $ out0_tb [3:0] $end
$var reg 16 % in_tb [15:0] $end
$scope module d0 $end
$var wire 16 & in [15:0] $end
$var wire 4 ' out3 [3:0] $end
$var wire 4 ( out2 [3:0] $end
$var wire 4 ) out1 [3:0] $end
$var wire 4 * out0 [3:0] $end
$upscope $end
$scope task verify $end
$var reg 16 + in_task [15:0] $end
$var reg 4 , out0_task [3:0] $end
$var reg 4 - out1_task [3:0] $end
$var reg 4 . out2_task [3:0] $end
$var reg 4 / out3_task [3:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b1100 /
b11 .
b1111 -
b1111 ,
b1100001111111111 +
b1111 *
b1111 )
b11 (
b1100 '
b1100001111111111 &
b1100001111111111 %
b1111 $
b1111 #
b11 "
b1100 !
$end
#1
b0 /
b0 .
b0 -
b0 ,
#102
