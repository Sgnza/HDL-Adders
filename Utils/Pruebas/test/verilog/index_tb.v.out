#! /c/Source/iverilog-install/bin/vvp
:ivl_version "12.0 (devel)" "(s20150603-1539-g2693dd32b)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "C:\iverilog\lib\ivl\system.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_sys.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_textio.vpi";
:vpi_module "C:\iverilog\lib\ivl\v2005_math.vpi";
:vpi_module "C:\iverilog\lib\ivl\va_math.vpi";
S_00000205a20f6380 .scope module, "index_tb" "index_tb" 2 2;
 .timescale 0 0;
v00000205a20e77c0_0 .var "in_tb", 15 0;
v00000205a20e7860_0 .net "out0_tb", 3 0, L_00000205a20e7ae0;  1 drivers
v00000205a20e7900_0 .net "out1_tb", 3 0, L_00000205a2148ff0;  1 drivers
v00000205a20e79a0_0 .net "out2_tb", 3 0, L_00000205a2149d60;  1 drivers
v00000205a20e7a40_0 .net "out3_tb", 3 0, L_00000205a2149b80;  1 drivers
S_00000205a20f6510 .scope module, "d0" "index" 2 11, 3 1 0, S_00000205a20f6380;
 .timescale 0 0;
    .port_info 0 /INPUT 16 "in";
    .port_info 1 /OUTPUT 4 "out0";
    .port_info 2 /OUTPUT 4 "out1";
    .port_info 3 /OUTPUT 4 "out2";
    .port_info 4 /OUTPUT 4 "out3";
v00000205a1e6bd70_0 .net "in", 15 0, v00000205a20e77c0_0;  1 drivers
v00000205a20f47e0_0 .net "out0", 3 0, L_00000205a20e7ae0;  alias, 1 drivers
v00000205a20c2d00_0 .net "out1", 3 0, L_00000205a2148ff0;  alias, 1 drivers
v00000205a1e6bae0_0 .net "out2", 3 0, L_00000205a2149d60;  alias, 1 drivers
v00000205a1e6b340_0 .net "out3", 3 0, L_00000205a2149b80;  alias, 1 drivers
L_00000205a20e7ae0 .part v00000205a20e77c0_0, 0, 4;
L_00000205a2148ff0 .part v00000205a20e77c0_0, 4, 4;
L_00000205a2149d60 .part v00000205a20e77c0_0, 8, 4;
L_00000205a2149b80 .part v00000205a20e77c0_0, 12, 4;
S_00000205a20e5090 .scope task, "verify" "verify" 2 19, 2 19 0, S_00000205a20f6380;
 .timescale 0 0;
v00000205a20f66a0_0 .var "in_task", 15 0;
v00000205a20f6740_0 .var "out0_task", 3 0;
v00000205a20e5220_0 .var "out1_task", 3 0;
v00000205a20e52c0_0 .var "out2_task", 3 0;
v00000205a20e5360_0 .var "out3_task", 3 0;
TD_index_tb.verify ;
    %load/vec4 v00000205a20f66a0_0;
    %store/vec4 v00000205a20e77c0_0, 0, 16;
    %delay 1, 0;
    %load/vec4 v00000205a20f6740_0;
    %load/vec4 v00000205a20e7860_0;
    %cmp/e;
    %flag_get/vec4 4;
    %jmp/0 T_0.3, 4;
    %load/vec4 v00000205a20e5220_0;
    %load/vec4 v00000205a20e7900_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_0.3;
    %flag_set/vec4 9;
    %flag_get/vec4 9;
    %jmp/0 T_0.2, 9;
    %load/vec4 v00000205a20e52c0_0;
    %load/vec4 v00000205a20e79a0_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_0.2;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.0, 8;
    %vpi_call 2 30 "$display", "CORRECTO: in: %b, out0 obtenido: %b, out1 obtenido: %b, out2 obtenido: %b, out3 obtenido: %b", v00000205a20e77c0_0, v00000205a20e7860_0, v00000205a20e7900_0, v00000205a20e79a0_0, v00000205a20e7a40_0 {0 0 0};
    %jmp T_0.1;
T_0.0 ;
    %vpi_call 2 32 "$display", "INCORRECTO: in: %b, out0 obtenido: %b, out1 obtenido: %b, out2 obtenido: %b, out3 obtenido: %b", v00000205a20e77c0_0, v00000205a20e7860_0, v00000205a20e7900_0, v00000205a20e79a0_0, v00000205a20e7a40_0 {0 0 0};
T_0.1 ;
    %end;
    .scope S_00000205a20f6380;
T_1 ;
    %vpi_call 2 39 "$dumpfile", "index_tb.vcd" {0 0 0};
    %vpi_call 2 40 "$dumpvars", 32'sb00000000000000000000000000000000, S_00000205a20f6380 {0 0 0};
    %end;
    .thread T_1;
    .scope S_00000205a20f6380;
T_2 ;
    %pushi/vec4 50175, 0, 16;
    %store/vec4 v00000205a20f66a0_0, 0, 16;
    %pushi/vec4 15, 0, 4;
    %store/vec4 v00000205a20f6740_0, 0, 4;
    %pushi/vec4 15, 0, 4;
    %store/vec4 v00000205a20e5220_0, 0, 4;
    %pushi/vec4 3, 0, 4;
    %store/vec4 v00000205a20e52c0_0, 0, 4;
    %pushi/vec4 12, 0, 4;
    %store/vec4 v00000205a20e5360_0, 0, 4;
    %fork TD_index_tb.verify, S_00000205a20e5090;
    %join;
    %pushi/vec4 50175, 0, 16;
    %store/vec4 v00000205a20f66a0_0, 0, 16;
    %pushi/vec4 0, 0, 4;
    %store/vec4 v00000205a20f6740_0, 0, 4;
    %pushi/vec4 0, 0, 4;
    %store/vec4 v00000205a20e5220_0, 0, 4;
    %pushi/vec4 0, 0, 4;
    %store/vec4 v00000205a20e52c0_0, 0, 4;
    %pushi/vec4 0, 0, 4;
    %store/vec4 v00000205a20e5360_0, 0, 4;
    %fork TD_index_tb.verify, S_00000205a20e5090;
    %join;
    %delay 100, 0;
    %vpi_call 2 62 "$finish" {0 0 0};
    %end;
    .thread T_2;
# The file index is used to find the file name in the following table.
:file_names 4;
    "N/A";
    "<interactive>";
    "index_tb.v";
    "./../../src/verilog/index.v";
